README.md
# Python bindings for Sveriges Radio API

Tested with Python version 2.5, 2.6, 2.7, 3.2 and 3.3.

## Installation

You can easily install sverigesradio from PyPi.

```bash
pip install sverigesradio
```

## Issues and Contribution

Bug reports are done by [creating an issue on Github](https://github.com/tiwilliam/sverigesradio/issues). If you want to contribute you can always [create a pull request](https://github.com/tiwilliam/sverigesradio/pulls) for discussion and code submission.

## Getting Started

SverigesRadio provides a simple interface for retrieving channels, programs, songs and more from Sveriges Radio's public API.

```python
>>> from sverigesradio import SverigesRadio

>>> SverigesRadio.channels()
[Channel(P1), Channel(P2), Channel(P3), Channel(P4 Blekinge), Channel(P4 Dalarna), Channel(P4 Gotland), Channel(P4 Gävleborg), Channel(P4 Göteborg), Channel(P4 Halland), Channel(P4 Jämtland), Channel(P4 Jönköping), Channel(P4 Kalmar), Channel(P4 Kristianstad), Channel(P4 Kronoberg), Channel(P4 Malmöhus), Channel(P4 Norrbotten), Channel(P4 Sjuhärad), Channel(P4 Skaraborg), Channel(P4 Stockholm), Channel(P4 Sörmland), Channel(P4 Uppland), Channel(P4 Värmland), Channel(P4 Väst), Channel(P4 Västerbotten), Channel(P4 Västernorrland), Channel(P4 Västmanland), Channel(P4 Örebro), Channel(P4 Östergötland), Channel(SR Sápmi), Channel(SR Sisuradio), Channel(P6), Channel(Din gata), Channel(Metropol), Channel(P2), Channel(P3 Star), Channel(Radioapans knattekanal), Channel(P4 Radiosporten), Channel(P2 Klassiskt), Channel(Minnen), Channel(P2 Världen), Channel(SR Extra01), Channel(SR Extra02), Channel(SR Extra03), Channel(SR Extra04), Channel(SR Extra05), Channel(SR Extra06), Channel(SR Extra07), Channel(SR Extra08), Channel(SR Extra09), Channel(SR Extra10), Channel(P2 Klassisk Jul), Channel(P4 Bjällerklang)]

>>> p3 = SverigesRadio.channel(164)
>>> p3.song.now
Song(Mat Zo & Porter Robinson - Easy)

>>> p3.program.now
Episode(Musikguiden i P3)

>>> p3.program.next
Episode(Ekonyheter)
```


CHANGELOG.md
# ChangeLog for sverigesradio

### Release 0.3

* New: First release

Uploaded to PyPi on 2013-12-20


sverigesradio/__init__.py
# -*- coding: utf-8 -*-

import re
import json
import requests
from datetime import datetime


class Song(object):
    def __init__(self, artist=None, title=None, description=None,
                 composer=None, conductor=None, **kwargs):
        self.artist = artist.encode('utf-8')
        self.title = title.encode('utf-8')
        self.description = description.encode('utf-8')
        self.composer = composer.encode('utf-8')
        self.conductor = conductor.encode('utf-8')

    def __repr__(self):
        return 'Song(%s - %s)' % (self.artist, self.title)


class Episode(object):
    def __init__(self, title=None, starttimeutc=None,
                 endtimeutc=None, **kwargs):
        self.title = title.encode('utf-8')
        self.starttimeutc = starttimeutc
        self.endtimeutc = endtimeutc

    def __repr__(self):
        return 'Episode(%s)' % self.title

    @staticmethod
    def json_to_datetime(date):
        match = re.match('/Date\((\d+)\)/', date)
        if not match:
            return None
        return datetime.fromtimestamp(int(match.group(1)) / 1000.0)

    @property
    def starttime(self):
        return self.json_to_datetime(self.starttimeutc)

    @property
    def endtime(self):
        return self.json_to_datetime(self.endtimeutc)


class Playlist(object):
    def __init__(self, now=None, next=None, content=Song):
        self.now = content(**now) if now else None
        self.next = content(**next) if next else None

    def __repr__(self):
        return 'Playlist(%s, %s)' % (self.now, self.next)


class Channel(object):
    def __init__(self, name=None, id=None, siteurl=None, color=None,
                 image=None, url=None, **kwargs):
        if not id:
            raise Exception('No such channel')

        self.id = id
        self.name = name.encode('utf-8')
        self.siteurl = siteurl
        self.color = color
        self.image = image
        self.url = url

    def __repr__(self):
        return 'Channel(%s)' % self.name

    @property
    def song(self):
        payload = dict(channelid=self.id)
        response = SverigesRadio.call('/playlists/rightnow', payload)
        if not response:
            return Playlist()

        playlist = response.get('playlist')
        if not playlist:
            return Playlist()

        song = playlist.get('song')
        nextsong = playlist.get('nextsong')

        return Playlist(now=song, next=nextsong)

    @property
    def program(self):
        payload = dict(channelid=self.id)
        response = SverigesRadio.call('/scheduledepisodes/rightnow', payload)
        if not response:
            return Playlist()

        channel = response.get('channel')
        if not channel:
            return Playlist()

        currentepisode = channel.get('currentscheduledepisode')
        nextepisode = channel.get('nextscheduledepisode')

        return Playlist(now=currentepisode, next=nextepisode, content=Episode)


class SverigesRadio(object):
    @classmethod
    def call(cls, method, payload={}):
        url = 'http://api.sr.se/api/v2/%s' % method

        payload['format'] = 'json'

        response = requests.get(url, params=payload)
        if response.status_code != 200:
            return {}

        return json.loads(response.text)

    @classmethod
    def channels(cls):
        payload = dict(size=500)
        data = cls.call('/channels', payload)
        channels = data.get('channels')
        return [Channel(**channel) for channel in channels]

    @classmethod
    def channel(cls, id):
        data = cls.call('/channels/%s' % id)
        return Channel(**data.get('channel', {}))

    @classmethod
    def schedule(cls, channelid=None, programid=None):
        payload = dict(size=500, channelid=channelid, programid=programid)
        data = cls.call('/scheduledepisodes', payload)
        if not data:
            return []
        schedule = data.get('schedule')
        return [Episode(**episode) for episode in schedule]


sverigesradio.py
# -*- coding: utf-8 -*-

import re
import json
import requests
from datetime import datetime

"""

SverigesRadio.channel(164).name
SverigesRadio.channel(164).song
SverigesRadio.channel(164).song.now
SverigesRadio.channel(164).song.next
SverigesRadio.channel(164).program

"""


class Song(object):
    def __init__(self, artist=None, title=None, description=None,
                 composer=None, conductor=None, **kwargs):
        self.artist = artist
        self.title = title
        self.description = description
        self.composer = composer
        self.conductor = conductor

    def __repr__(self):
        return 'Song(%s - %s)' % (self.artist, self.title)


class Episode(object):
    def __init__(self, title=None, starttimeutc=None,
                 endtimeutc=None, **kwargs):
        self.title = title
        self.starttimeutc = starttimeutc
        self.endtimeutc = endtimeutc

    def __repr__(self):
        return 'Episode(%s)' % self.title

    @staticmethod
    def json_to_datetime(date):
        match = re.match('/Date\((\d+)\)/', date)
        if not match:
            return None
        return datetime.fromtimestamp(int(match.group(1)) / 1000.0)

    @property
    def starttime(self):
        return self.json_to_datetime(self.starttimeutc)

    @property
    def endtime(self):
        return self.json_to_datetime(self.endtimeutc)


class Playlist(object):
    def __init__(self, now=None, next=None, content=Song):
        self.now = content(**now) if now else None
        self.next = content(**next) if next else None

    def __repr__(self):
        return 'Playlist(%s, %s)' % (self.now, self.next)


class Channel(object):
    def __init__(self, name=None, id=None, siteurl=None, color=None,
                 image=None, url=None, **kwargs):
        if not id:
            raise Exception('No such channel')

        self.id = id
        self.name = name
        self.siteurl = siteurl
        self.color = color
        self.image = image
        self.url = url

    def __repr__(self):
        return 'Channel(%s)' % self.id

    @property
    def song(self):
        payload = dict(channelid=self.id)
        response = SverigesRadio.call('/playlists/rightnow', payload)
        if not response:
            return Playlist()

        playlist = response.get('playlist')
        if not playlist:
            return Playlist()

        song = playlist.get('song')
        nextsong = playlist.get('nextsong')

        return Playlist(now=song, next=nextsong)

    @property
    def program(self):
        payload = dict(channelid=self.id)
        response = SverigesRadio.call('/scheduledepisodes/rightnow', payload)
        if not response:
            return Playlist()

        channel = response.get('channel')
        if not channel:
            return Playlist()

        currentepisode = channel.get('currentscheduledepisode')
        nextepisode = channel.get('nextscheduledepisode')

        return Playlist(now=currentepisode, next=nextepisode, content=Episode)


class SverigesRadio(object):
    @classmethod
    def call(cls, method, payload={}):
        url = 'http://api.sr.se/api/v2/%s' % method

        payload['format'] = 'json'

        response = requests.get(url, params=payload)
        if response.status_code != 200:
            return {}

        return json.loads(response.text)

    @classmethod
    def channels(cls):
        payload = dict(size=500)
        data = cls.call('/channels', payload)
        channels = data.get('channels')
        return [Channel(**channel) for channel in channels]

    @classmethod
    def channel(cls, id):
        data = cls.call('/channels/%s' % id)
        return Channel(**data.get('channel', {}))

    @classmethod
    def schedule(cls, channelid=None, programid=None):
        payload = dict(size=500, channelid=channelid, programid=programid)
        data = cls.call('/scheduledepisodes', payload)
        if not data:
            return []
        schedule = data.get('schedule')
        return [Episode(**episode) for episode in schedule]
